nautobot_circuit_maintenance/jobs/site_search.py

```diff
@@ -0,0 +1,161 @@
+    """
+
+    start: datetime
+    end: datetime
+
+
+def check_for_overlap(record1: CircuitMaintenance, record2: CircuitMaintenance):
+    """Checks for the overlap of two circuit maintenance records.
+
+    Args:
+        record1 (CircuitMaintenance): First maintenance record
+        record2 (CircuitMaintenance): Second maintenance record
+
+    Returns:
+        bool: Result of there is overlap
+    """
+    # Build a couple of ranges
+    range1 = Range(start=record1.start_time, end=record1.end_time)
+    range2 = Range(start=record2.start_time, end=record2.end_time)
+
+    # Determine the latest start and the earliest end to determine overlap
+    latest_start = max(range1.start, range2.start)
+    earliest_end = min(range1.end, range2.end)
+
+    # Get the delta, compare versus 0 (negative number is no overlap)
+    delta = (earliest_end - latest_start).days + 1
+
+    # If delta is negative there is no overlap, return false
+    overlap = max(0, delta)
+
+    return overlap != 0
+
+
+def get_sites_from_circuit(circuit: Circuit):
+    """Get a list of sites from a circuit object in Nautobot.
+
+    Args:
+        circuit (Circuit): Nautobot Circuit Object to find the sites
+
+
+    Returns:
+        list: List of Nautobot Site objects
+    """
+    site_list: list = []
+    for term in [circuit.termination_a, circuit.termination_z]:
+        if term is not None and getattr(term, "provider_network") is None and term.site not in site_list:
+            site_list.append(term.site)
+
+    return site_list
+
+
+def build_sites_to_maintenance_mapper(maintenance_queryset):
+    """Build a site to circuit maintenance mapper so the data can be quickly accessed of what possible overlaps.
+
+    Leverages defaultdict to provide the default value of an empty set to each key that will be added. Then adds each
+    particular circuit maintenance object to the dictionary to be used as a map of maintenances going on at the site.
+    Build a data dictionary that maps sites to circuit maintenances:
+
+    data_dict = {
+        "msp": {
+            CM1,
+            CM2
+        },
+        "nyc": {
+            CM3,
+            CM4
+        }
+    }
+
+    Args:
+        maintenance_queryset (Queryset): Queryset of all Circuit Maintenance objects
+
+    Returns:
+        dict: Dictionary of a set of maintenance records
+    """
+    return_dictionary = collections.defaultdict(set)
+    for record in maintenance_queryset:
+        # Get the site overall
+        # Start with looping over the circuits that are defined
+        for circuit in record.circuits:
+            # Check both termination A and Z for values of None, add that record to the set for each site
+            for term in [circuit.termination_a, circuit.termination_z]:
+                if term is not None:
+                    return_dictionary[term.site.name].add(record)
+
+    return dict(return_dictionary)
+
+
+class FindSitesWithCircuitImpact(Job):
```

Comment:
CircuitImpact has already a specific meaning in this plugin, maybe another name would make it clearer